File: leetcode/medium/921/count.cpp
/**
 * @file stack.cpp
 * @author Vortexx2
 * @brief Problem 921 - Minimum Add To Make Parentheses Valid
 * @approach Arithmetic checks and balances
 * @date 30-06-2021
 * 
 * Runtime - 0ms
 * Memory Usage - 6.1 MB
 */
#include <iostream>
#include <stack>
#include <string>

using namespace std;

class Solution
{
public:
  int minAddToMakeValid(string s)
  {
    int bal = 0, count = 0;

    for (auto &ch : s)
    {
      bal += ((ch == '(') ? 1 : -1);

      if (bal == -1)
      {
        bal = 0;
        count += 1;
      }
    }
    return bal + count;
  }
};

File: leetcode/medium/921/count.py
"""
  @author Vortexx2
  Problem 921 - Minimum Add to Make Parentheses Valid
  @approach Maintaining a balance for the open parentheses

  Runtime - 24 ms
  Memory Usage - 14.4 MB
"""

class Solution:
  def minAddToMakeValid(self, s: str) -> int:

    bal = count = 0

    for ch in s:

      bal += 1 if ch == '(' else -1

      if bal == -1:
        bal = 0
        count += 1

    return bal + count
